target/linux/mediatek/filogic/base-files/lib/upgrade/platform.sh

```diff
@@ -1,5 +1,129 @@
 REQUIRE_IMAGE_METADATA=1
 
+redmi_ax6000_nand_upgrade_tar()
+{
+	CI_UBIPART=ubi
+	local tar_file="$1"
+	local board_dir="$(tar tf "$tar_file" | grep -m 1 '^sysupgrade-.*/$')"
+	board_dir="${board_dir%/}"
+
+	local kernel_length=$( (tar xf "$tar_file" "$board_dir/kernel" -O | wc -c) 2> /dev/null)
+	[ "$kernel_length" = 0 ] && kernel_length=
+
+	local rootfs_length=$( (tar xf "$tar_file" "$board_dir/root" -O | wc -c) 2> /dev/null)
+	[ "$rootfs_length" = 0 ] && rootfs_length=
+
+	local rootfs_type
+	[ "$rootfs_length" ] && rootfs_type="$(identify_tar "$tar_file" "$board_dir/root")"
+
+	[ -n "$rootfs_length" -o -n "$kernel_length" ] || return 1
+
+	local mtdnum="$( find_mtd_index "$CI_UBIPART" )"
+	if [ ! "$mtdnum" ]; then
+		echo "cannot find ubi mtd partition ubi"
+		return 1
+	fi
+	local ubidev="$( nand_find_ubi "$CI_UBIPART" )"
+	#cleanup old data volume if exist
+	if [ "$ubidev" ] && [ "$( nand_find_volume $ubidev data )" ]; then
```

Comment:
@dangowrt  we have to cleanup old `data` volume if exist, this is diff from `nand_upgrade_tar`

Reply:
I understand. Please add this functionality to `nand_upgrade_tar` and have it guarded by a new environment variable, such as `$CI_KERN_UBIPART`.

Reply:
@dangowrt but how to handle the action like `delete existing volume data`?

Reply:
You can either do that with yet another additional env variable which is then checked by in `nand_upgrade_tar`, but as it is very specific for this board and only needs to be done *once* (ie. first update after switching to OpenWrt, coming from vendor stock ROM, and not every time OpenWrt is updated after that) that may be overkill.

A better option than duplicating all of `nand_upgrade_tar` (I really want to avoid that), is maybe to create a wrapper-function which takes care of the removal of that `data` volume and then calls `nand_upgrade_tar "$@"`.

Reply:
it maybe a big change on these functions `nand_upgrade_tar` and `nand_upgrade_prepare_ubi` 
I think I cannot handle that.

Reply:
and I don't believe there will be a second device that will use these features

Reply:
More and more vendors start to use UBI to load kernel, but then choose a not very advantageous layout (e.g. for dual-boot or for other unknown reasons) with very small UBI partition to hold their firmware. So I believe this will actually be needed in future if we want to stay compatible with those vendor decisions (currently we advice users to flash initramfs with different flash layout and then do sysupgrade from there, making it hard to revert to stock, just like for this device).